pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableDebugResource.java

```diff
@@ -0,0 +1,441 @@
+    try {
+      switch (tableType) {
+        case OFFLINE:
+          return TableIngestionStatusHelper
+              .getOfflineTableIngestionStatus(tableNameWithType, _pinotHelixResourceManager,
+                  _pinotHelixTaskResourceManager);
+        case REALTIME:
+          return TableIngestionStatusHelper.getRealtimeTableIngestionStatus(tableNameWithType,
+              _controllerConf.getServerAdminRequestTimeoutSeconds() * 1000, _executor, _connectionManager,
+              _pinotHelixResourceManager);
+        default:
+          break;
+      }
+    } catch (Exception e) {
+      return TableStatus.IngestionStatus.newIngestionStatus(TableStatus.IngestionState.UNKNOWN, e.getMessage());
+    }
+    return null;
+  }
+
+  private TableDebugInfo.TableSizeSummary getTableSize(String tableNameWithType) {
+    TableSizeReader tableSizeReader =
+        new TableSizeReader(_executor, _connectionManager, _controllerMetrics, _pinotHelixResourceManager);
+    TableSizeReader.TableSizeDetails tableSizeDetails;
+    try {
+      tableSizeDetails = tableSizeReader
+          .getTableSizeDetails(tableNameWithType, _controllerConf.getServerAdminRequestTimeoutSeconds() * 1000);
+    } catch (Throwable t) {
+      tableSizeDetails = null;
+    }
+
+    return (tableSizeDetails != null) ? new TableDebugInfo.TableSizeSummary(tableSizeDetails._reportedSizeInBytes,
+        tableSizeDetails._estimatedSizeInBytes) : new TableDebugInfo.TableSizeSummary(-1, -1);
+  }
+
+  /**
+   * Helper method to debug segments. Computes differences between ideal state and external view for each segment.
+   *
+   * @param pinotHelixResourceManager Helix Resource Manager
+   * @param tableNameWithType Name of table with type
+   * @param verbosity Verbosity level to include debug information. For level 0, only segments
+   *                  with errors are included. For level > 0, all segments are included.   * @return Debug
+   *                  information for segments
+   */
+  private List<TableDebugInfo.SegmentDebugInfo> debugSegments(PinotHelixResourceManager pinotHelixResourceManager,
+      String tableNameWithType, int verbosity) {
+    ExternalView externalView = pinotHelixResourceManager.getTableExternalView(tableNameWithType);
+    IdealState idealState = pinotHelixResourceManager.getTableIdealState(tableNameWithType);
+
+    List<TableDebugInfo.SegmentDebugInfo> result = new ArrayList<>();
+    if (idealState == null) {
+      return result;
+    }
+
+    int serverRequestTimeoutMs = _controllerConf.getServerAdminRequestTimeoutSeconds() * 1000;
+    final Map<String, List<String>> serverToSegments =
+        _pinotHelixResourceManager.getServerToSegmentsMap(tableNameWithType);
+
+    BiMap<String, String> serverToEndpoints;
+    try {
+      serverToEndpoints = _pinotHelixResourceManager.getDataInstanceAdminEndpoints(serverToSegments.keySet());
+    } catch (InvalidConfigException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)